clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1201,11 +1250,71 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
   }
 };
 
+// A type to check the validity of passing union with accessor/sampler/stream
+// member as a kernel argument types.
+class SyclKernelUnionBodyChecker : public SyclKernelFieldHandler {
+  static constexpr const bool VisitUnionBody = true;
+  int UnionCount = 0;
+  bool IsInvalid = false;
+  DiagnosticsEngine &Diag;
+
+public:
+  SyclKernelUnionBodyChecker(Sema &S)
+      : SyclKernelFieldHandler(S), Diag(S.getASTContext().getDiagnostics()) {}
+  bool isValid() { return !IsInvalid; }
+
+  bool enterUnion(const CXXRecordDecl *RD, FieldDecl *FD) {
+    ++UnionCount;
+    return true;
+  }
+
+  bool leaveUnion(const CXXRecordDecl *RD, FieldDecl *FD) {
+    --UnionCount;
+    return true;
+  }
+
+  bool handlePointerType(FieldDecl *FD, QualType FieldTy) final {
+    if (UnionCount) {
+      IsInvalid = true;
+      Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+          << FieldTy;
+    }
+    return isValid();
+  }
+
+  bool handleSyclAccessorType(FieldDecl *FD, QualType FieldTy) final {
+    if (UnionCount) {
+      IsInvalid = true;
+      Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+          << FieldTy;
+    }
+    return isValid();
+  }
+
+  bool handleSyclSamplerType(FieldDecl *FD, QualType FieldTy) final {
+    if (UnionCount) {
+      IsInvalid = true;
+      Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+          << FieldTy;
+    }
+    return isValid();
+  }
+  bool handleSyclStreamType(FieldDecl *FD, QualType FieldTy) final {
+    if (UnionCount) {
+      IsInvalid = true;
+      Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+          << FieldTy;
+    }
+    return isValid();
+  }
+};
+
 // A type to Create and own the FunctionDecl for the kernel.
 class SyclKernelDeclCreator : public SyclKernelFieldHandler {
   FunctionDecl *KernelDecl;
   llvm::SmallVector<ParmVarDecl *, 8> Params;
   SyclKernelFieldChecker &ArgChecker;
+  SyclKernelUnionBodyChecker &ArgChecker1;
```

Comment:
I'd probably call this 'UnionChecker'.  Additionally, note that #2289 makes this part unnecessary.

Reply:
Ok. I will change that.

Reply:
Actually, looking again, adding ArgChecker1 is completely unnecessary.  ArgChecker itself is completely unnecessary here too (and likely should be removed).  I remove it in #2289 (depending on when that gets in), but it seems to be a leftover from when it was necessary.

Reply:
It seems like only place we can add this handler  "SyclKernelUnionBodyChecker checker1(*this)" to ConstructOpenCLKernel.

Reply:
Yep, I thought you would have noticed trying to debug why diagnostics weren't working right, but it needs to be added to these two lines instead:

  Visitor.VisitRecordBases(KernelObj, checker, kernel_decl, kernel_body,
                           int_header);


  Visitor.VisitRecordFields(KernelObj, checker, kernel_decl, kernel_body,
                            int_header);